File: leetcode/algorithms/linked-list/remove-nth-node-from-end-of-list.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Remove Nth Node From End of List"
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        ListNode primaryNode=head,scoutNode=null;
        while(primaryNode!=null){
            scoutNode=primaryNode;
            int i=0;
            while(i<=n && scoutNode!=null){
                scoutNode=scoutNode.next;
                ++i;
            }
            if(i<=n){
                head = head.next;
                break;
            }
            if(scoutNode==null){
                primaryNode.next = primaryNode.next.next;
                break;
            }
            primaryNode=primaryNode.next;
        }
        return head;
        
        
    }
}

File: leetcode/algorithms/linked-list/remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

 function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    
    if(head===null){
        return head;
    }
    
    let i: number = 0,ptr: ListNode = head;
    
    while(i<n){
        ptr = ptr.next;
        ++i;
    }
    
    let prevNode: ListNode = null, nextNode: ListNode = head;
    
    while(ptr!==null){
        prevNode = nextNode;
        nextNode = nextNode.next;
        ptr = ptr.next;
    }
    
    if(prevNode===null){
        head = head.next;
    } else {
        prevNode.next = prevNode.next.next;
    }
    
    return head;

};